From d7922a0ae9a621a0be728f134953d6bafc412d0d Mon Sep 17 00:00:00 2001
From: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>
Date: Mon, 13 May 2024 22:03:35 -0700
Subject: [PATCH] wifi: ath12k: fix sta bandwidth to wmi channel width
 conversion logic

Currently "ath12k_mac_ieee80211_sta_bw_to_wmi" function converts the
station bandwidth to corresponding WMI channel width based on the
ieee80211_sta's deflink's bandwidth instead of the link specific
bandwidth. This results in sending wrong bandwidth value to the Firmware.

Fix this issue by passing the link specific bandwidth value for conversion

Patch-dependency: 766-ath12k-fix-firmware-assert-during-channel-switch-for.patch

Signed-off-by: Ganesh Kariganuru Mahabalesh <quic_gkarigan@quicinc.com>

Index: backports-6.5-rc3/drivers/net/wireless/ath/ath12k/mac.c
===================================================================
--- backports-6.5-rc3.orig/drivers/net/wireless/ath/ath12k/mac.c
+++ backports-6.5-rc3/drivers/net/wireless/ath/ath12k/mac.c
@@ -7465,11 +7465,11 @@ exit:
 }
 
 static u32 ath12k_mac_ieee80211_sta_bw_to_wmi(struct ath12k *ar,
-					      struct ieee80211_sta *sta)
+					      struct ieee80211_link_sta *link_sta)
 {
 	u32 bw = WMI_PEER_CHWIDTH_20MHZ;
 
-	switch (sta->deflink.bandwidth) {
+	switch (link_sta->bandwidth) {
 	case IEEE80211_STA_RX_BW_20:
 		bw = WMI_PEER_CHWIDTH_20MHZ;
 		break;
@@ -7487,7 +7487,7 @@ static u32 ath12k_mac_ieee80211_sta_bw_t
 		break;
 	default:
 		ath12k_warn(ar->ab, "Invalid bandwidth %d in rc update for %pM\n",
-			    sta->deflink.bandwidth, sta->addr);
+			    link_sta->bandwidth, link_sta->addr);
 		bw = WMI_PEER_CHWIDTH_20MHZ;
 		break;
 	}
@@ -8209,7 +8209,7 @@ static void ath12k_mac_op_sta_rc_update(
 	spin_lock_bh(&ar->data_lock);
 
 	if (changed & IEEE80211_RC_BW_CHANGED) {
-		bw = ath12k_mac_ieee80211_sta_bw_to_wmi(ar, sta);
+		bw = ath12k_mac_ieee80211_sta_bw_to_wmi(ar, link_sta);
 		arsta->bw_prev = arsta->bw;
 		arsta->bw = bw;
 	}
